addons/udes_stock/models/stock_move_line.py

```diff
@@ -621,3 +623,72 @@ def _split_and_group_mls_by_quantity(self, maximum_qty):
             remaining_mls -= mls_used
             grouped_mls.append(mls_used)
         return grouped_mls
+
+    def _merge_move_lines(self):
+        """ This method will, for each move line in `self`, go up in their linked picking and try to
+        find in their existing move line a candidate into which we can merge the move.
+        :return: Recordset of move lines passed to this method.
+        If some of passed move lines were merged into another existing one, return this one and
+        not the (now unlinked) original.
+        """
+        MoveLine = self.env["stock.move.line"]
+        distinct_fields = self._prepare_merge_move_lines_distinct_fields()
+
+        candidate_move_line_list = []
+        for picking in self.mapped("picking_id"):
+            candidate_move_line_list.append(picking.move_line_ids)
+        # Move removed after merge
+        move_lines_to_unlink = MoveLine.browse()
+        move_lines_to_merge = []
+        for candidate_move_lines in candidate_move_line_list:
+            # First step find move lines to merge.
+            candidate_move_lines = candidate_move_lines.with_context(prefetch_fields=False)
+            for k, g in groupby(sorted(
+                    candidate_move_lines, key=self._prepare_merge_move_line_sort_method
+            ), key=itemgetter(*distinct_fields)):
+                move_lines = MoveLine.concat(*g).filtered(
+                    lambda m: m.state not in ("done", "cancel")
+                )
+                # If we have multiple records we will merge then in a single one.
+                if len(move_lines) > 1:
+                    move_lines_to_merge.append(move_lines)
+
+        # second step merge its move lines, initial demand, ...
+        for move_lines in move_lines_to_merge:
+            # merge move line data
+            move_lines[0].write(move_lines._merge_move_lines_fields())
+            # update merged moves dicts
+            move_lines_to_unlink |= move_lines[1:]
+
+        if move_lines_to_unlink:
+            move_lines_to_unlink.sudo().unlink()
+        return (self | MoveLine.concat(*move_lines_to_merge)) - move_lines_to_unlink
+
+    @api.model
+    def _prepare_merge_move_lines_distinct_fields(self):
+        """
+        Prepare merge move lines distinct fields
+        """
+        return [
+            "product_id", "location_id", "location_dest_id", "package_id", "lot_id", "state"
```

Comment:
Do we need state? all move lines of the picking should be in the same state.

Do we want move_id? we could have 2 move lines with the same info but different move_id, so they cannot be merged.

Reply:
Added move_id as suggested, added even _merge_moves because noticed that was created a new move for same product. Can you have a look please?